### PR TITLE
fix(provider): resolve Tiingo security currency from countryCode

### PR DESCRIPTION
**Pull request by jestinjoshi**
## Summary
- Tiingo's `/tiingo/utilities/search` response never includes the `priceCurrency` field the code was reading, so currency detection silently failed for every security.
- Currency is now derived from `countryCode` via the `countries` gem's ISO 4217 data (already used the same way in `Provider::TwelveData`), instead of a hand-maintained allowlist.
- Added a `best_match_for_ticker` tie-break for the case where Tiingo's search returns multiple entries sharing the same ticker across countries (e.g. a US primary listing and a foreign cross-listing), so the currency shown in search results always matches what `fetch_security_prices` later returns for that ticker.

## Test plan
- [x] `bin/rails test test/models/provider/tiingo_test.rb` (10/10 passing)
- [x] `bin/rubocop app/models/provider/tiingo.rb test/models/provider/tiingo_test.rb` (no offenses)
- [x] `bin/brakeman --no-pager` (0 warnings)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved Tiingo currency resolution by deriving currency from each ticker’s best-matching listing country code.
  * Correctly handles duplicate ticker matches by prioritizing US listings to keep USD-based results consistent.
  * Expands currency mapping for additional country codes while preventing incorrect USD defaults for unrecognized codes.
  * Avoids overwriting an existing cached USD currency with a non-US-derived value.
  * Ensures failed currency determination returns an explicit error instead of a fallback.

* **Tests**
  * Added comprehensive provider tests covering currency resolution, caching behavior, and request strategy for price fetching.

### CODE REVIEW

File: app/models/provider/tiingo.rb [hunk 1]
```diff
@@ -280,6 +289,31 @@ def map_exchange_to_mic(exchange_name)
...
     end
 
+    # Tiingo's search/utilities response never includes a priceCurrency field
+    # (confirmed against the live API), only countryCode. Resolve the currency
+    # via the countries gem's ISO 4217 data (already used for country
+    # resolution in Provider::TwelveData) instead of hand-maintaining a
+    # per-provider allowlist.
+    def currency_for_country(country_code)
+      return nil if country_code.blank?
+      ISO3166::Country.new(country_code.strip)&.currency_code
```
[the remark this answers is not in the document]
Comment (jestinjoshi) [reply app/models/provider/tiingo.rb:315]: Checked this against the actual pinned version — `countries` 8.0.3 (see `Gemfile.lock`). `ISO3166::Country.new` is overridden in [`class_methods.rb`](https://github.com/countries/countries/blob/master/lib/countries/country/class_methods.rb):

```ruby
def new(country_data)
  return super if country_data.is_a?(Hash)
  country_code = country_data.to_s
  country_code = country_code.upcase if country_code.match?(/[a-z]/)
  super if codes.include?(country_code)  # implicit nil when code is unrecognized
end
```

For an unrecognized code, `super` (the real allocating `new`) is never called and the method falls through to `nil` — confirmed directly:

```
$ ruby -e 'require "countries"; p ISO3166::Country.new("ZZ")'
nil
```

So `ISO3166::Country.new(country_code.strip)&.currency_code` never reaches `currency_code` on invalid data — `&.` short-circuits on a true nil. This is also exercised end-to-end by the `"ZZ"` case in `tiingo_test.rb`, which passes with the intended `nil` currency, not an error.

Might be describing a different gem version's behavior — happy to add a test pinning this contract if useful insurance against a future `countries` upgrade changing it, but no code change needed for the currently pinned version.

File: app/models/provider/tiingo.rb [hunk 2]
```diff
@@ -83,15 +83,24 @@ def search_securities(symbol, country_code: nil, exchange_operating_mic: nil)
...
+      # per unique ticker and reuse it both for caching (so fetch_security_prices
+      # can use it without a second search request) and for the Security objects
+      # below, so what's shown in search results always matches what
+      # fetch_security_prices will later return.
+      currency_by_ticker = parsed.filter_map { |security| security["ticker"] }.map(&:upcase).uniq.index_with do |ticker|
+        currency_for_country(best_match_for_ticker(parsed, ticker)&.dig("countryCode"))
+      end
+
+      currency_by_ticker.each do |ticker, currency|
+        Rails.cache.write("tiingo:currency:#{ticker}", currency, expires_in: 24.hours) if currency.present?
```
Comment (jjmata) [inline app/models/provider/tiingo.rb]: Worth double-checking: this cache write is now unconditional and keyed only by ticker (`"tiingo:currency:#{ticker}"`), with no check for an existing value. Under the old code this was effectively dead — `priceCurrency` was never present in Tiingo's response, so the `if currency.present?` guard was always false and nothing was ever cached. Now that a real currency is computed here, every `search_securities` call for a ticker will overwrite whatever was previously cached for it.

`best_match_for_ticker` derives that currency only from the entries present in the *current* `parsed` array (i.e. whatever this particular query string happened to return). If a later search query for the same ticker returns a response that doesn't include the US listing (e.g. Tiingo's relevance ranking for a different query surfaces only the foreign cross-listing), `matches.find { us } || matches.first` falls back to that foreign entry and silently overwrites a previously-correct USD value with the wrong currency — which `fetch_security_prices` would then pick up from cache for a US-priced instrument.

This seems like a real (if narrow) staleness risk given the cache has no "don't downgrade a US-derived currency" guard, and I don't see a test exercising two searches for the same ticker where the second search's result set differs from the first (e.g. only returns a non-US entry). Not necessarily blocking, but might be worth either guarding the overwrite or adding a regression test for it.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by jjmata — COMMENTED** [review]
Reviewed the diff for `app/models/provider/tiingo.rb` and `test/models/provider/tiingo_test.rb` against the `Provider::TwelveData` precedent.

**Dependency**: `countries` is already in `Gemfile` (`gem "countries"`) and pinned in `Gemfile.lock` (`countries (8.0.3)`), used elsewhere (e.g. `Provider::TwelveData#search_securities` via `ISO3166::Country.find_country_by_any_name`). Confirmed this is not a new dependency — no CLAUDE.md "minimize dependencies" concern. (Small nuance: TwelveData actually uses the gem only for country *name* resolution, not currency derivation — it reads `currency` straight from the TwelveData API. That doesn't make this PR's approach wrong, just worth noting the "used the same way" framing in the PR description is a little generous.)

**`best_match_for_ticker`**: handles the edge cases I checked correctly — blank ticker → `nil`, no matches → `nil` (safe, `Array#find`/`first` don't raise on empty), all-non-US matches → falls back to `matches.first`, multiple US matches → picks the first (a reasonable tie-break given Tiingo's daily-price endpoints are ticker-only and effectively US-centric per the PR's live-API notes).

**Unrecognized `countryCode`**: confirmed `ISO3166::Country.new("ZZ").currency_code` resolves to `nil` rather than raising or defaulting, so `currency_for_country` correctly returns `nil` instead of silently defaulting to USD, and `fetch_currency_for_symbol` raises `Error` rather than mislabeling. This is exercised by the "unrecognized country code" tests in both `search_securities` and `fetch_security_prices`.

**Test coverage**: good breadth — US/no-`priceCurrency`, CA via countryCode, a country outside the old allowlist (FR), unrecognized countryCode (both search and fetch-prices paths), the AAPL duplicate-ticker (CA+US) tie-break for both display and caching, and the cache-hit vs. cache-miss-fallback paths for `fetch_security_prices`. Verified the `Rails.cache.stubs(...)`-with-`MemoryStore` pattern is necessary and accurate since `config.cache_store = :null_store` in `config/environments/test.rb`.

Left one comment on a possible staleness edge case in the per-ticker cache write (line 98) — the cache is now unconditionally overwritten on every search using only the current query's result set, so a later search that doesn't happen to return the US entry for an already-cached ticker could downgrade a correct USD value to a foreign currency. Not necessarily blocking, but doesn't look like it's covered by a test.

Overall this is a solid fix for a real bug (the old code silently never worked since `priceCurrency` doesn't exist in Tiingo's response) with much better test coverage than before.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: app/models/provider/tiingo.rb [hunk 3]
```diff
@@ -83,15 +83,40 @@ def search_securities(symbol, country_code: nil, exchange_operating_mic: nil)
...
+      # per unique ticker and reuse it both for caching (so fetch_security_prices
+      # can use it without a second search request) and for the Security objects
+      # below, so what's shown in search results always matches what
+      # fetch_security_prices will later return.
+      matches_by_ticker = parsed.filter_map { |security| security["ticker"] }.map(&:upcase).uniq.index_with do |ticker|
+        best_match_for_ticker(parsed, ticker)
+      end
+
+      currency_by_ticker = matches_by_ticker.transform_values { |match| currency_for_country(match&.dig("countryCode")) }
+
```
Comment (jestinjoshi) [inline app/models/provider/tiingo.rb:98]: Fixed in c55766fd — the cache write in `search_securities` now only overwrites an already-cached currency when the current search's result set has a US match for that ticker (or nothing is cached yet), so a later search that happens not to surface the US cross-listing can no longer downgrade a previously-cached USD. Added a regression test (`search_securities does not downgrade a cached US-derived currency when a later search for the same ticker omits the US entry`) exercising exactly the scenario you described.

🤖 Generated with [Claude Code](https://claude.ai/code)

**Review by jjmata — APPROVED** [review]
